File: social-flow-backend/shared/libraries/python/security/rbac.py
# rbac.py
from typing import Dict, List


class RBAC:
    """
    Role-Based Access Control.
    """

    def __init__(self):
        self.roles: Dict[str, List[str]] = {}

    def add_role(self, role: str, permissions: List[str]):
        self.roles[role] = permissions

    def check_permission(self, role: str, permission: str) -> bool:
        return permission in self.roles.get(role, [])

File: social-flow-backend/shared/libraries/node/auth/rbac.js
/**
 * Role-Based Access Control (RBAC)
 */

class RBAC {
  constructor() {
    this.roles = new Map(); // role -> permissions
  }

  addRole(role, permissions = []) {
    this.roles.set(role, new Set(permissions));
  }

  grant(role, permission) {
    if (!this.roles.has(role)) this.roles.set(role, new Set());
    this.roles.get(role).add(permission);
  }

  revoke(role, permission) {
    if (this.roles.has(role)) this.roles.get(role).delete(permission);
  }

  can(role, permission) {
    return this.roles.has(role) && this.roles.get(role).has(permission);
  }
}

const rbac = new RBAC();

// Default roles
rbac.addRole('user', ['read:self']);
rbac.addRole('admin', ['read:any', 'write:any', 'delete:any']);

module.exports = rbac;
